pkgs/development/compilers/ghc/8.2.1.nix

```diff
@@ -32,33 +61,111 @@ in stdenv.mkDerivation (rec {
+    gmp.dev
+  ] ++ stdenv.lib.optionals (!useVendoredLibffi) [
+    libffi.dev
+  ];
+
+  buildInputs = stdenv.lib.optionals (targetPlatform == hostPlatform) [
+    alex autoconf automake ghc happy hscolour perl python3 sphinx
+  ] ++ stdenv.lib.optionals (targetPlatform != hostPlatform) [
+    targetStdenv.cc
+
+    # Stringly speaking, LLVM is only needed for platforms the native
+    # code generator does not support, but using it when
+    # cross-compiling anywhere.
+    llvmPackages.llvm
+
+    ncurses.dev __targetPackages.ncurses.dev
+  ] ++ stdenv.lib.optionals (!enableIntegerSimple && targetPlatform != hostPlatform) [
+    gmp.dev __targetPackages.gmp.dev
+  ] ++ stdenv.lib.optionals (!useVendoredLibffi) [
+    libffi.dev __targetPackages.libffi.dev
+  ];
+
+  propagatedBuildInputs = stdenv.lib.optionals (targetPlatform != hostPlatform) [
+    ncurses.out __targetPackages.ncurses.out
+  ] ++ stdenv.lib.optionals (!enableIntegerSimple && targetPlatform != hostPlatform) [
+    gmp.out __targetPackages.gmp.out
+  ] ++ stdenv.lib.optionals (!useVendoredLibffi) [
+    libffi.out __targetPackages.libffi.out
+  ] ++ stdenv.lib.optionals (targetPlatform != hostPlatform && targetPlatform.libc == "libsystem") [
+    __targetPackages.libiconv
+  ];
 
-  buildInputs = commonBuildInputs;
+  # Hack so that envHooks are run for build inputs and crossEnvHooks for host
+  # inputs. In the future stdenv's setup.sh will be changed to avoid this need.
+  crossConfig = if targetPlatform != hostPlatform then "hack" else null;
 
   enableParallelBuilding = true;
 
   configureFlags = [
-    "CC=${stdenv.cc}/bin/cc"
-    "--with-curses-includes=${ncurses.dev}/include" "--with-curses-libraries=${ncurses.out}/lib"
+    "CC=${targetStdenv.cc or stdenv.cc}/bin/${prefix}cc"
+  # TODO: With Cross rebuild (need to fix hooks) remove these `--with-*` altogether
+  ] ++ stdenv.lib.optionals (targetPlatform == hostPlatform) [
+    "--with-curses-includes=${__targetPackages.ncurses.dev}/include"
+    "--with-curses-libraries=${__targetPackages.ncurses.out}/lib"
+  ] ++ [
     "--datadir=$doc/share/doc/ghc"
-  ] ++ stdenv.lib.optional (! enableIntegerSimple) [
-    "--with-gmp-includes=${gmp.dev}/include" "--with-gmp-libraries=${gmp.out}/lib"
-  ] ++ stdenv.lib.optional stdenv.isDarwin [
-    "--with-iconv-includes=${libiconv}/include" "--with-iconv-libraries=${libiconv}/lib"
+  ] ++ stdenv.lib.optionals (targetPlatform == hostPlatform && !enableIntegerSimple) [
+    "--with-gmp-includes=${__targetPackages.gmp.dev}/include"
+    "--with-gmp-libraries=${__targetPackages.gmp.out}/lib"
+  ] ++ stdenv.lib.optionals (targetPlatform == hostPlatform && targetPlatform.isDarwin) [
+    "--with-iconv-includes=${__targetPackages.libiconv}/include"
+    "--with-iconv-libraries=${__targetPackages.libiconv}/lib"
+  ] ++ stdenv.lib.optionals (targetPlatform != hostPlatform) [
+
+    # TODO: next rebuild make these unconditional
```

Comment:
Just open this against staging and do it now IMO

Reply:
Done.